### ISSUE TITLE
Proposal: `tail return` recursive calls

### ISSUE THREAD

**Issue by agocke** [dotnet/csharplang#2304]
There are a number of instances where it may be clearer to express your algorithm via recursion, especially tail recursion, but unfortunately all recursion in C# currently requires O(n) space for just the call stack, meaning that your algorithm cannot take less than O(n) space.

It is proposed that C# provide a new statement form and a new expression form to allow tail-recursive calls to consume O(1) space.

## Syntax and Semantics

The statement form would be `tail return <expr>`, analogous to `yield return`, where the `expr` target is required and must be a tail recursive call.

The expression form would be `tail <expr>`, where the `<expr>` would have the same restrictions as the statement form, and the `tail <expr>` expression would only be legal as the expression in an expression-bodied method.

## Codegen

There is a `tail` instruction in the CLR, but if that instruction is not reliable or desired, the compiler can lower this into a `goto` after assigning the new arguments for the recursive call.

For instance,

```C#
void M(int x)
{
    ...
    tail return M(x - 1);
}
```

Could be rewritten to

```C#
void M(int x)
{
    start:
    ...
    x = x - 1;
    goto start;
}
```

### Limitations

Because the compiler cannot reliably rewrite inter-procedural calls, this would only currently be legal for tail-recursive calls. It would not be allowed for e.g., mutually tail recursive calls. If the CLR can reliably perform `tail` calls without significant downsides on all supported platforms, the feature could be broadened from the current specification.

**Comment by agocke** [comment]
cc @gafter

**Comment by YairHalberstadt** [comment]
Has any analysis been done to see how many locations in eg. Roslyn could benefit from this?

Is a new keyword necessary, or would an attribute be sufficient, as this doesn't change the semantics of the program?

**Comment by YairHalberstadt** [comment]
Would rewriting mutually recursive local functions be considered?

**Comment by CyrusNajmabadi** [comment]
So, presumably you can write;

```c#
if (x)
{
    tail return MyMethod(a, b, c);
}
else
{
    tail return MyMethod(d, e, f)
}
```

But would you be able to do:

```c#
tail return x ? MyMethod(a, b, c) : MyMethod(d, e, f);
```

?

**Comment by agocke** [comment]
@CyrusNajmabadi Not as written, no.

**Comment by agocke** [comment]
@YairHalberstadt Regardless of what we could do for local functions, the restrictions as listed would not allow mutual tail recursion.

As for attributes, the important part is really the call, not the definition, right? We can't put attributes on expressions, so if the call is really the part we want to annotate, attributes are not suitable.

**Comment by CyrusNajmabadi** [comment]
@agocke Understood.  Probably not critical to support.  but it would be nice if it was relaxed to teh generalized definition of tail-calls. Namely that they must be the last thing the function executes.  In such a formalization, conditionals would be fine.  

Just for my own edification, would this be challenging to support?

**Comment by agocke** [comment]
Probably not. The nice thing about tail calls is you don't really care about all the stuff nested inside the expression, you only need to know that 1) the outer expression is a call and 2) the call is recursive.

For conditional, that would just add an introspection "through" the conditional to both sides. From a language point of view it's a bit unfortunate to dig through expressions, but it's probably necessary for the broader semantics to work properly.

**Comment by CyrusNajmabadi** [comment]
> For conditional, that would just add an introspection "through" the conditional to both sides. From a language point of view it's a bit unfortunate to dig through expressions, but it's probably necessary for the broader semantics to work properly with conditional.

Makes sense.  Thanks for clarifying!

**Comment by HaloFour** [comment]
It seems a little weird trying to generate explicit tail call behavior in the C# language given that IL supports it and RyuJIT supports it as an optimization implicitly.  I can understand the concerns with the reliability of those features and explicit language support would make it more predictable.

Given that tail recursion and functional programming seem to go hand-in-hand I'd want to ensure that a feature like this would cover those common bases, like being embedded in a `switch` expression.  My (limited) experience with `@tailrec` in Scala is that it's often somewhat tricky/obnoxious to find the right order of operations to satisfy the compiler.

**Comment by CyrusNajmabadi** [comment]
One more question:

What if you do:

```c#
void M(int x)
{
    ...
    tail return other.M(x - 1);
}
```

Seems like this should be supported.  This would just be equiv to slamming over the 'this' param with 'other' and then tailing, right?  

Or will this only be allowed for static methods, or tail-methods with 'this' as the receiver?  Thanks!

**Comment by CyrusNajmabadi** [comment]
> It seems a little weird trying to generate explicit tail call behavior in the C# language

I actually don't have a problem with this.  It's less about the compiler actually generating something appropraite, and it's more about actually ensuring you either get tail calls, or you fail to compile.  For example, in my programming life i've made *constant* mistake around code i thought should be tail-called.  It is so simple to do something accidental that makes tail calls stop working, and you can end up shooting yourself in the foot.

So, i see the presence of `tail` in the language much more of a check to say "you better not have screwed this up!". 

It's similar to a `static` local function for me.  Yes, i can just not capture anything.  But `static` makes sure that if i mess that up in the future it just won't compile.

**Comment by HaloFour** [comment]
I know that C# does not yet have list patterns, but something like this would be critical to support:

```cs
public int Sum(int seed, List<int> list) => list switch {
    [value ... rest] => tail Sum(seed + value, rest),
    _ => seed
};
```

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> It's less about the compiler actually generating something appropraite, and it's more about actually ensuring you either get tail calls, or you fail to compile.

This proposal seems to be about emitting the explicit `goto` IL to implement the tail call and skipping whatever IL/JIT support there is.  I'd agree that I'd probably prefer a `tail` operator (or attribute+analyzer) to simply enforce what the runtime supports, but that's probably a difficult target to nail down.  IIRC the runtime is free to ignore the `tail.` opcode prefix, and implicit support varies quite a bit.

**Comment by CyrusNajmabadi** [comment]
>  but something like this would be critical to support:

I would just like the proposal relaxed to "return points".  Then, it should "just work" for conditional-exprs, switch-statements (though that likely doesn't need anything special), switch-exprs, expr-bodies, and anything else that has that concept.

**Comment by agocke** [comment]
@HaloFour Good point. That would not work with this proposal but it does seem unfortunate that it wouldn't.

That's really the same problem as conditional, where making this work requires digging through the semantics of the `switch`, which is a little gross 😕

**Comment by HaloFour** [comment]
Scala example:

```scala
@tailrec
private def sumWithAccumulator(list: List[Int], accumulator: Int): Int = {
    list match {
        case Nil => accumulator
        case x :: xs => sumWithAccumulator(xs, accumulator + x)
    }
}
```

**Comment by CyrusNajmabadi** [comment]
>This proposal seems to be about emitting the explicit goto IL to implement the tail call and skipping whatever IL/JIT support there is.

I'm breaking it apart in my head.  First, i'm just assessing what the language feature is.  And i think it would be valuable to have this language feature (for the reason i specified above).  If the language feature is good, it's definitely worth discussing what the right emit-strategy is.  Honestly, given how simple the 'goto' form should be, i really would have no problem with the langauge just spitting that out.  However, if there's high confidence that the .tail instruction works for the entirety of the ecosystem, i think it's fine if that's the emit strategy.

**Comment by CyrusNajmabadi** [comment]
> which is a little gross 

From a language perspective, it seems utterly natural to me :)    Or do you mean 'gross' in the sense of the impl?

**Comment by agocke** [comment]
>First, i'm just assessing what the language feature is.

This proposal is primarily about getting guaranteed O(1) behavior for tail call recursion. 

Any codegen strategy that can meet that requirement is up for grabs.

**Comment by CyrusNajmabadi** [comment]
Minor point, but syntactically, we could just consider this being a 'tail expression' and then only allowing the 'tail expression' in certain constructs.

that way, you would write it more like this `return a ? tail Foo() : tail Foo()`.  You would also do `Foo() => x switch { y => tail Foo() ...`.

In this formalization, expression-bodies are not special cased.  They just naturally fall out.  

Note: this is how we handled 'ref' as well.  i.e. we don't have a ref-conditional internally.  We just allow 'ref expressions' and those expressions are only allowed in special constructs.  From an impl perspective, this also makes it much easier to do error tolerance.

**Comment by CyrusNajmabadi** [comment]
Note: the approach as currently stated seems like it would not allow using 'tail' for void methods.  Maybe that's not a big deal, but it seems a bit odd.  If this is relaxed to "the last op must be a call to the enclosing method", then this can support tail recursive void methods fine, where the last call is just a recursive call to yourself.

**Comment by agocke** [comment]
> Note: this is how we handled 'ref' as well. i.e. we don't have a ref-conditional internally. We just allow 'ref expressions' and those expressions are only allowed in special constructs. 

Nope, that's not how it works. In the language we created new special "ref" variants of each of those language forms and then special-cased *those* behaviors, creating an incredibly complex ref matrix. You're thinking of how we implemented it in the compiler, where there is such a thing as a "ref expression" (which implements said incredibly complicated ref matrix 😄).

**Comment by HaloFour** [comment]
@CyrusNajmabadi 

> If the language feature is good, it's definitely worth discussing what the right emit-strategy is. Honestly, given how simple the 'goto' form should be, i really would have no problem with the langauge just spitting that out. However, if there's high confidence that the .tail instruction works for the entirety of the ecosystem, i think it's fine if that's the emit strategy.

I agree, it's worth exploring.  It's just weird to me that we're adding a third mechanism for tail recursion.  But maybe the third time is the charm?  😁 

@agocke 

In these cases would we want to allow the same method to invoke itself normally?  If not, why not take a page from Scala and make it a modifier of the method itself?  Or an attribute that the compiler recognizes?  That would involve no changes to the grammar of the language, I think.

I wouldn't have much of an opinion one way or the other, I'm only making the comparison with Scala.  I can see a benefit to a compiler-enforced modifier/attribute that ensures that the entire method supports tail recursion.  But I can also see an advantage to the escape hatch to support normal recursion in an alternate code path.

Also, I think it's worthwhile to bring in the F# folks and discuss the scenarios in which that compiler will emit the `tail.` IL prefix, their experiences as to when the runtime actually supports it and whether similar `goto`-based codegen was discussed.

**Comment by agocke** [comment]
@HaloFour I see no reason why you couldn't mix tail recursive calls and regular recursive calls.

**Comment by HaloFour** [comment]
I didn't realize that the F# compiler also rewrites the function into a loop.  I thought they emitted `tail.` and let the runtime optimize it.  Perhaps they had the same experience of it being unreliable?

```fsharp
let rec sum values accum =
    match values with
    | [] -> accum
    | head :: tail -> sum tail (accum + head)
```

https://sharplab.io/#v2:DYLgZgzgNALiCWwA+wCmMAEAnVBjDEArgLYYBuAhsIahBhbriRgLwCwAUBtxsRTLgAW5KjToB3eDEGceGJBgDaAXQwBaAHz1GJWTwWDUFACYYQIDDAqJ1WoqSs2AFAyakA1BkMmAlJyA

**Comment by CyrusNajmabadi** [comment]
> Nope, that's not how it works. In the language we created new special "ref" variants of each of those language forms and then special-cased those behaviors, creating an incredibly complex ref matrix. You're thinking of how we implemented it in the compiler, where there is such a thing as a "ref expression" (which implements said incredibly complicated ref matrix 😄).

We originally considered entirely different syntactic forms here where the 'ref' wouldn't necessarily even be in a position that it could be interpretted to be part of hte expression.  We then with a language form that had the benefit that you could think of it either as a specialized form, or as a 'ref expression'.  We went with that because of the attractive nature at the grammar and the impl level.

I'm arguing we should do the same.  Instead of thnking about 'tail' as being related to 'return', think about 'tail' as being something you do with a *call*.  That is both a nicer way to conceptualize it (imo) *and* it leads to a nicer implementation.  It's win/win for me.

**Comment by YairHalberstadt** [comment]
I think it's probably useful for proposals like this to have a number of real life examples of cases where it might be applied, so that the LDM can be sure they've considered all potential use cases.

Here is an example of a method I'm writing which both uses void tail recursion, and mixes tail recursive and non tail recursive calls:

```csharp
nternal static void VisitRankSpecifiers<TArg>(this TypeSyntax type, Action<ArrayRankSpecifierSyntax, TArg> action, TArg argument)
        {
            switch (type)
            {
                case ArrayTypeSyntax arrayTypeSyntax:
                    arrayTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
                    foreach (var rankSpecifier in arrayTypeSyntax.RankSpecifiers)
                    {
                        action(rankSpecifier, argument);
                    }
                    break;
                case NullableTypeSyntax nullableTypeSyntax:
                    nullableTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
                    break;
                case PointerTypeSyntax pointerTypeSyntax:
                    pointerTypeSyntax.ElementType.VisitRankSpecifiers(action, argument);
                    break;
                case TupleTypeSyntax tupleTypeSyntax:
                    foreach (var element in tupleTypeSyntax.Elements)
                    {
                        element.Type.VisitRankSpecifiers(action, argument);
                    }
                    break;
                case RefTypeSyntax refTypeSyntax:
                    refTypeSyntax.Type.VisitRankSpecifiers(action, argument);
                    break;
                case GenericNameSyntax genericNameSyntax:
                    foreach (var typeArgument in genericNameSyntax.TypeArgumentList.Arguments)
                    {
                        typeArgument.VisitRankSpecifiers(action, argument);
                    }
                    break;
                case QualifiedNameSyntax qualifiedNameSyntax:
                    qualifiedNameSyntax.Left.VisitRankSpecifiers(action, argument);
                    qualifiedNameSyntax.Right.VisitRankSpecifiers(action, argument);
                    break;
                case IdentifierNameSyntax _:
                    break;
                case OmittedTypeArgumentSyntax _:
                    break;
                case PredefinedTypeSyntax _:
                    break;
                case AliasQualifiedNameSyntax _:
                    break;
                default:
                    throw ExceptionUtilities.Unreachable;
            }
```

An extra optimization that tail recursion would make possible, would be to avoid copying `argument` on every call, which is helpful for large structs. I imagine though that is an implementation detail, and wouldn't be guaranteed by the spec.

**Comment by orthoxerox** [comment]
Obligatory reference to an earlier discussion: https://github.com/dotnet/roslyn/issues/1235

While I agree that interprocedural tail recursion and tail recursion modulo cons would be great, I'd rather the LDM hammer out the simplest case first: tail call support in non-virtual self-recursive methods. That alone would solve like 80% of all cases.

**Comment by PathogenDavid** [comment]
How would the `goto` codegen interact with `stackalloc`? As far as I can find, there's no way to explicitly free memory allocated with the `localloc` instruction.

On the `tail.` side of things, according to the spec, it can't be used inside of any exception blocks.

**Comment by jnm2** [comment]
@CyrusNajmabadi 

> I'm arguing we should do the same. Instead of thnking about 'tail' as being related to 'return', think about 'tail' as being something you do with a _call_. That is both a nicer way to conceptualize it (imo) _and_ it leads to a nicer implementation. It's win/win for me.

This makes complete sense to me.

Also, there are times I want a tail call when I can't use the `return` keyword because the return type of the current method and the called method are both `void`.

**Comment by johnkellyoxford** [comment]
You can simply guarantee the O(1) recursion size by translating it to a loop

**Comment by gafter** [comment]
@agocke The standard computer-science meaning of a [tail call](https://en.wikipedia.org/wiki/Tail_call) is a call which is the final action.  It is not necessary for it to be a self-call to be a tail call.  The last "Limitations" paragraph of the OP appears to assume "tail call" means "tail self call".  It is probably worth clarifying earlier in the proposal that this is only intended to support self tail calls.

Many of the places where I would love to use tail calls in the Roslyn compilers are in the visitors.  Many of the methods on the call stack in the binder or flow analysis are simple methods that end in a (not self) tail call, where it would be valuable to discard the stack frame.

**Comment by agocke** [comment]
@gafter That's true. That's what I meant by "recursive" call. It's technically true that recursion does not require *direct* recursion, but I was trying to keep it simple

**Comment by samuel-vidal** [comment]
I think it is **an excellent idea**

_With C# supporting a very confortable style of functional programming it becomes increasingly valuable to have a construct to help guaranteeing stack safety._

If we agree this is the purpose of that proposal, I think we should precise its semantic as follows:

The proposed 'tail' syntax should mean exactly that : the compiler is asked to check that the call will be optimized by the Jit and emit the necessary IL.

In my opinion it would be preferable to make the IL .tail work than to make it a compiler code rewrite (this is way too narrow, and way more complex).

A tail call is one that is performed just before exiting a method. Either returning (nothing) directly if that call returns void or return the result of that call in case it returns something.
No operation on the return value should be performed (not even implicit cast, although upcast should be fine).
A key point is that for a call to be tail, both methods have to share the same return type (can be relaxed a bit with covariance).

Also, _having the criteria implemented in the compiler_, the .tail IL should be emitted even in the absence of the tail constraint. 

proposed syntax :
`
tail qualifiedmethodname(parameters);	// (the syntax is not tied to the return statement).
`

examples

```c#
void method1()
{
	tail method1();		// tail call
	return;
	...
	method2();		// tail call
	return;
}

void method2()
{
	method1();		// tail call
	return;
	...
	method3();		// Not tail
	return;
}

int method3()
{
	method1();		// Not tail
	return 1;
	...
	return method3();		// tail call
	...
	return method4() + 1;		// Not tail
}

int method4()
{
	return tail method3();		// tail call
	...
	tail method3();		// compilation fails with an error (or a warning)
	return 1;
	...
	retrun x == j ? tail method3() : method4() + 1;	// only the first call is tail.
}
```

A few points:

The real difficulty here, I think, will be the interaction with ref/in/out arguments (let alone stack alloc etc) rather than the tail call being to the same method.
A point to have in mind is that a virtual method calling itself is actually performing another virtual dispatch (could be confusing for the user).

**Scope** : The compilation should fail or produce a warning if the condition for guaranteeing tail call optimization by the Jit are not match(or too hard to figure out e.g. in the case of in/out/ref args etc).

**Comment by johnkellyoxford** [comment]
>  the compiler is asked to check that the call will be optimized by the Jit

compiler can't check that. most it can do is emit `.tail` prefix

**Comment by CyrusNajmabadi** [comment]
> compiler can't check that. most it can do is emit .tail prefix

Compiler can also literally just do the tail recursion itself.  Which would actually be my preference.   I wouldn't want to ever do `tail return` but then end up at runtime stack overflowing.

**Comment by johnkellyoxford** [comment]
yeah, but it cannot check the JIT optimization or JIT behaviour in any way

**Comment by johnkellyoxford** [comment]
what do you mean by "do the tail recursion itself" tho - the `.tail` prefix?

**Comment by gafter** [comment]
> Compiler can also literally just do the tail recursion itself

Not without changing the signature of methods in the case of mutual recursion.

**Comment by CyrusNajmabadi** [comment]
> what do you mean by "do the tail recursion itself" tho - the .tail prefix?

Actually performing the optimization in the method.  i.e. not calling the recursive method, but instead just converting it to a goto directly.

**Comment by CyrusNajmabadi** [comment]
> Not without changing the signature of methods in the case of mutual recursion.

i believe andy's proposal put that out of scope.  but i could be mistaken.  I thought this only covered self-recursion.

**Comment by samuel-vidal** [comment]
Maybe I should make a separate proposal (sorry new here) since The design I would favor, would be to strengthen the contract with the runtime in a way that all CLR language could benefit from.
(sorry new here: would you suggest me to make a separate proposal or this discussion is a fine place).

**Comment by CyrusNajmabadi** [comment]
No worries.  Given your interest, i do personally think a coreclr issue would be more relevant if this is about ensuring behavior around the `.tail` instruction.

**Comment by samuel-vidal** [comment]
Also, I should not have been that specific. I could change:

> The proposed 'tail' syntax should mean exactly that : the compiler is asked to check that the call will be optimized by the Jit and emit the necessary IL.

to

> The proposed 'tail' syntax should mean exactly that : the compiler is asked to check that 1) the call is tail and 2) guarantee that the call will be optimized (either by the Jit through emitting the necessary IL or otherwise).

**Comment by johnkellyoxford** [comment]
@samuel-vidal 
![image](https://user-images.githubusercontent.com/32420055/57966933-7e9be000-7950-11e9-95a8-659a02c218b1.png)

That contract already exists

**Comment by johnkellyoxford** [comment]
Also, that is not the runtime @samuel-vidal , that is just the compiler. There are various *different* runtimes and JITs

> The proposed 'tail' syntax should mean exactly that : the compiler is asked to check that 1) the call is tail and 2) guarantee that the call will be optimized (either by the Jit through emitting the necessary IL or otherwise).

Compiler cannot check JIT behaviour as i said before. It can ONLY:
* Emit the `tail.` prefix itself, which guarantees desired behaviour according to ECMA 335
* Emit the tail call itself, transforming it to a loop

**Comment by samuel-vidal** [comment]
@johnbeisner I think that is fair enough. Since the contract exists and is guaranteed to be honored by the runtime then point 2) is simply emitting tail. prefix.

I was under the impression that the tail call optimization is subject to a long list of Byzantine conditions in order to have the actual desired effect. So it would be valuable to have the compiler warn the user if any of those conditions are not met. I came accross [that list](https://blogs.msdn.microsoft.com/davbr/2007/06/20/tail-call-jit-conditions/) for example (nb it is quite old and I have no idea how relevant it still is).

**Comment by samuel-vidal** [comment]
Also interaction with `try/catch` is another can of worms, a `return tail call(param)` within a `try` is not exactly a tail call, you can't erase all the stack state for the caller method. (so the conditions you noted above can't be met). That would have to be one of the case when the compiler says no : 'Tail constraints within a try block are not allowed'.

**Comment by Thaina** [comment]
Totally +1

I would also wish that the compiler will just error when we put `tail` into any line that not possible to be tail called

**Comment by hez2010** [comment]
IIRC tail return is guaranteed by JIT: https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8BLAOwwAIBZACjMoQEoBYAKAG82LuLiAzCjQQUAvCIoBGRhQACAdkkBuLj3nUhFOJMbLWAXyA==

**Comment by timcassell** [comment]
@hez2010 Maybe in simple cases like that, but what about with virtual and non-self recursion?

```cs
void Func()
{
    // ...
    FuncVirt();
}

protected virtual void FuncVirt()
{
    // ...
    Func2();
}

void Func2()
{
    // ...
    if (x)
    {
        Func();
    }
}
```
